src/main/java/org/broadinstitute/hellbender/tools/dataflow/transforms/markduplicates/MarkDuplicatesUtils.java

```diff
@@ -216,6 +266,209 @@ public void processElement(final ProcessContext context) throws Exception {
                 });
     }
 
+    // This transform takes marked reads and generates the metrics from those tests.
+    public static class GenerateMetricsTransform extends PTransform<PCollection<GATKRead>, PCollection<KV<String, DuplicationMetrics>>> {
+        private static final long serialVersionUID = 1l;
+
+        /**
+         *  The header for all reads processed.
+        */
+        private final PCollectionView<SAMFileHeader> header;
+
+        public GenerateMetricsTransform(final PCollectionView<SAMFileHeader> header) {
+            this.header = header;
+        }
+
+        @Override
+        public PCollection<KV<String, DuplicationMetrics>> apply(final PCollection<GATKRead> preads) {
+            return preads
+                .apply(generateMetricsByLibrary())
+                .setCoder(KvCoder.of(StringUtf8Coder.of(), new DuplicationMetricsCoder()))
+                // Combine metrics for reads from the same library.
+                .apply(Combine.<String, DuplicationMetrics>perKey(new CombineMetricsFn()))
```

Comment:
Is scaling an issue with his combiner?  I didn't notice before but it seems like it is making a metric per read and then combining all the metrics per library in a single combination.  Do you know if this is an issue or does it have magic to make this into a hierarchical combination that can be scattered?